include/boost/geometry/algorithms/detail/overlay/pointlike_linear.hpp

```diff
@@ -0,0 +1,344 @@
+            geometry::expand(total, geometry::return_envelope<Box>(geometry));
+        }
+
+    };
+
+    struct overlaps_box
+    {
+        template <typename Box, typename Geometry>
+        static inline bool apply(Box const& box, Geometry const& geometry)
+        {
+            return ! geometry::disjoint(geometry, box);
+        }
+    };
+
+    template <typename OutputIterator>
+    class item_visitor_type
+    {
+    public:
+        item_visitor_type(OutputIterator& oit) : m_oit(oit) {}
+
+        template <typename Item1, typename Item2>
+        inline void apply(Item1 const& item1, Item2 const& item2)
+        {
+            action_selector_pl_l
+                <
+                    PointOut, overlay_intersection
+                >::apply(item1, Policy::apply(item1, item2), m_oit);
+        }
+
+    private:
+        OutputIterator& m_oit;
+    };
+    // structs for partition -- end
+
+    class segment_range
+    {
+    public:
+        typedef geometry::segment_iterator<Linear const> const_iterator;
+        typedef const_iterator iterator;
+
+        segment_range(Linear const& linear)
+            : m_linear(linear)
+        {}
+
+        const_iterator begin() const
+        {
+            return geometry::segments_begin(m_linear);
+        }
+
+        const_iterator end() const
+        {
+            return geometry::segments_end(m_linear);
+        }
+
+    private:
+        Linear const& m_linear;
+    };
+
+    template <typename OutputIterator>
+    static inline OutputIterator get_common_points(MultiPoint const& multipoint,
+                                                   Linear const& linear,
+                                                   OutputIterator oit)
+    {
+        item_visitor_type<OutputIterator> item_visitor(oit);
+
+        segment_range rng(linear);
+
+        geometry::partition
+            <
+                geometry::model::box
+                    <
+                        typename boost::range_value<MultiPoint>::type
+                    >,
+                expand_box,
+                overlaps_box
+            >::apply(multipoint, rng, item_visitor);
+
+        return oit;
+    }
+
+public:
+    template <typename RobustPolicy, typename OutputIterator, typename Strategy>
+    static inline OutputIterator apply(MultiPoint const& multipoint,
+                                       Linear const& linear,
+                                       RobustPolicy const& robust_policy,
+                                       OutputIterator oit,
+                                       Strategy const& strategy)
+    {
+        typedef std::vector
+            <
+                typename boost::range_value<MultiPoint>::type
+            > point_vector_type;
+
+        point_vector_type common_points;
+
+        // compute the common points
+        get_common_points(multipoint, linear,
+                          std::back_inserter(common_points));
+
+        return multipoint_multipoint_point
```

Comment:
Yes, your understanding is correct: `common_points` contains all points of the `multipoint` that are in the intersection of `multipoint` and `linear`.

Let me start by saying that I do understand your concerns. I had the same questions/thoughts when implementing these set operations and finally decided to implement them as you see in this PR for the reasons described below.

Let's consider that this `apply()` method does: it is used to compute either the difference or the intersection of `multipoint` with `linear`. More analysis below.

#### Case 1: Intersection
I agree with you that in this case the visitor could do the job. However there is a draw back in this approach: if the multipoint contains a point `p` (and assume that `p` appears in `multipoint` once) that is contained in many segments of the linear geometry, then `p` will be part of the output multipoint many times (as many times as the number of segments in the linear geometry that contain it). I wanted to avoid that and maintain the property that a point of `multipoint`, that is part of the intersection of the multipoint and the linear geometry, appears in the output as many times as it appears in `multipoint`.

#### Case 2: Difference
Difference is a bit trickier: suppose that `multipoint` contains a point `p` such that it intersects a segment `s1` of `linear` but does not intersect another segment `s2` of `linear`. Then the visitor for partition will discard `p` when considered against `s1` but will keep it when considered against `s2`. So `p` will actually appear in the output which is wrong.

Essentially what I am saying is that in order to correctly implement `difference(multipoint, linear)` using `bg::partition` I first need to compute the intersection of `multipoint` and `linear` and then take the difference of `multipoint` and the intersection computed. This is essentially what the code does. On top of that, the current implementation for difference maintains the same property as intersection: a point of `multipoint` appears in the output (if is part of the output) with the same *multiplicity* as in the input `multipoint`.

#### Summary
The current approach has been used for two reasons:
* Algorithm correctness (this applies to the difference algorithm).
* For a point `p` in the input that is also part of the output, we maintain the invariant that the multiplicity of `p` in the input multipoint stays the same in the output multipoint.